### PR TITLE
[GSoC] Add `SearchNotes()` function

### PR DESCRIPTION
**Pull request by viciousAegis**
## Purpose / Description
Prep commit for notes browser

## Approach
Add a similar method to `searchCards()` that calls the `findNotes()` function from Collection.

## How Has This Been Tested?

Manually on my phone.

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration (SDK version(s), emulator or physical, etc)

## Learning (optional, can help others)
_Describe the research stage_

_Links to blog posts, patterns, libraries or addons used to solve this problem_

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by viciousAegis** [comment]
This is getting a bit slow on large collections (> 3 seconds on a 2000 note collection), and I think to make it faster we might have to write a direct database query. Any other suggestions & ideas appreciated.

Also, waiting to get tests up and running on my Mac to add unit tests

**Review by david-allison — APPROVED** [review]
Looks good to me

But to note: I dislike the concept of `CardCache`, and if we have time, we should refactor this.

**Comment by mikehardy** [comment]
I was thinking more in the direction that @dae has the current habit of popping by on these ones and saying "I've already implemented this for the new schema path and it's 1ms for 10k cards so no worries, just waiting on Mike to merge PR 3.1415" 😁

**Comment by david-allison** [comment]
It has, we're just not hooked up to it yet.

`backend.searchNotes`

it can be overridden:

https://github.com/ankidroid/Anki-Android/blob/231b847f08e3bd99ef748c8fcc355d0dd7065006/AnkiDroid/src/main/java/com/ichi2/libanki/CollectionV16.kt#L96-L118

The API will be the same, as we'd override `col.findNotes`

**Review by mikehardy — APPROVED** [review]
LGTM Thanks!

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt [hunk 1]
```diff
@@ -661,6 +661,46 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
...
+            val searchResult: MutableList<CardCache> = ArrayList()
+            val searchResult_: List<Long>
+            searchResult_ = try {
+                col.findNotes(query).requireNoNulls()
+            } catch (e: Exception) {
+                // exception can occur via normal operation
+                Timber.w(e)
+                return SearchCardsResult.error(e)
+            }
+            Timber.d("The search found %d cards", searchResult_.size)
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:680]: %d notes technically

**Review by Arthur-Milchior — COMMENTED** [review]
Do you think it could be possible to:
* reduce the amount of copy paste. Probably using an extra parameter to indicate whether you want note or card, and using an if/else in the small part of the code that changes.
* clean up the code (the IDE show multiple places that can automatically be corrected here)


Without looking at the profiler, I can guess that the part that is slow is `firstCard` because it requires to do one query for each note. And I really mean each note, even the one we decide not to display yet. I entirely agree with you, the simpler solution would be to change the query.

I can imagine we could create an interface out of CardCache, by considering only the part that is useful for the browser's work. If we do that, we can then ensure that the browser's displays expect this interface instead of the actual CardCache object. Then we could implement a similar NoteCache, reusing a lot of code in common. The main difference being that if we actually uses `firstCard` we can delay it until card content must be displayed. (Or ideally, never use it at all because we never require a use of the card itself)


But this seems to be a lot of work with very little gain here. Especially when changing the query is extremely simple. 
Do you know how to do the change, or do you want help ? Essentially, you can really easily re-use the code from `findNote`, and you only need to know about grouping-by note ids.

**Comment by david-allison** [comment]
The architecture of CardCache needs a lot of work. Let's not solidify it with a `NoteCache`

**Comment by dae** [comment]
Probably not worth trying to optimize the query, as AD will be able to use the backend version soon. Rather than use a cache, AD could likely achieve further speedups in the future by relying on the backend to render the browser rows.

**Comment by viciousAegis** [comment]
> Do you think it could be possible to:
> 
> reduce the amount of copy paste. Probably using an extra parameter to indicate whether you want note or card, and using an if/else in the small part of the code that changes.
> clean up the code (the IDE show multiple places that can automatically be corrected here)

I will be doing that in the coming commits for this. Here is the `Search` function that I am testing right now, I have modified `SearchCards` in  `CardBrowser.kt`. I have created a bool to keep track of whether the browser is in Cards mode or Notes mode.
The current function:
https://github.com/ankidroid/Anki-Android/blob/35f74f018e2e418e3e6f3285890c7f03ecfaf535/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt#L1479-L1511

My modified version with the `isCardsMode` boolean:

```kotlin
private fun searchCards() {
        // cancel the previous search & render tasks if still running
        invalidate()
        if (mSearchTerms == null) {
            mSearchTerms = ""
        }
        if ("" != mSearchTerms && mSearchView != null) {
            mSearchView!!.setQuery(mSearchTerms!!, false)
            mSearchItem!!.expandActionView()
        }
        val searchText: String? = if (mSearchTerms!!.contains("deck:")) {
            "($mSearchTerms)"
        } else {
            if ("" != mSearchTerms) "$mRestrictOnDeck($mSearchTerms)" else mRestrictOnDeck
        }

        if (colIsOpen() && mCardsAdapter != null) {
            mCards.reset()
            mCardsAdapter!!.notifyDataSetChanged()
            if (isCardsMode) {
                TaskManager.launchCollectionTask(
                    SearchCards(
                        searchText!!,
                        if (mOrder == CARD_ORDER_NONE) NoOrdering() else UseCollectionOrdering(),
                        numCardsToRender(),
                        mColumn1Index,
                        mColumn2Index
                    ),
                    mSearchCardsHandler
                )
            } else {
                TaskManager.launchCollectionTask(
                    SearchCards(
                        searchText!!,
                        if (mOrder == CARD_ORDER_NONE) NoOrdering() else UseCollectionOrdering(),
                        numCardsToRender(),
                        mColumn1Index,
                        mColumn2Index
                    ),
                    mSearchCardsHandler
                )
            }
        }
    }
```

This works like intended, and doesn't require significant changes to achieve that

**Pull request merged**